atintegrators/ExactHamiltonianPass.cc

```diff
@@ -0,0 +1,207 @@
+
+#define NUM_FIELDS_2_REMEMBER (sizeof(FieldNames) / sizeof(FieldNames[0]))
+
+void ExactHamiltonianPass(double *r_in, double le,
+                     double *A, double *B,
+                     const double *T1, const double *T2,
+                     const double *R1, const double *R2,
+                     int max_order, int num_int_steps,
+                     double phi, int type,
+                     double gK, int multipole_fringe,
+                     int num_particles)
+{
+    int c, n;
+    double * r6;
+
+    // copy AT structure into pass method structure
+
+    element e = {0};
+
+    for(n = 0; n < max_order; n++)
+    {
+        e.F[2 * n]     = B[n];
+        e.F[2 * n + 1] = A[n];
+    }
+    
+    e.L = le;
+    e.phi = phi;
+    e.gK = gK;
+    e.nF = max_order;
+    e.slices = num_int_steps;
+    e.type = type;
+    e.do_multipole_fringe = multipole_fringe;
+
+    for(c = 0; c<num_particles; c++)
+    {
+        r6 = r_in+c*6;	
+        if(!atIsNaN(r6[0]))
+        {   
+            /* misalignment at entrance */
+            if(T1)
+                ATaddvv(r6,T1);
+            if(R1)
+                ATmultmv(r6,R1);
+            
+            track_element(r6, &e);
+            
+            /* misalignment at exit */
+            if(R2)
+                ATmultmv(r6,R2);
+            if(T2)   
+                ATaddvv(r6,T2);
+        }
+    }
+}
+
+#if defined(MATLAB_MEX_FILE) || defined(PYAT)
+
+ExportMode struct elem *trackFunction(const atElem *ElemData, struct elem *Elem,
+                               double *r_in, int num_particles,
+                               struct parameters *Param)
+
+{
+    double le, bending_angle;
+    double *polynom_a, *polynom_b;
+    long max_order, num_int_steps, type, multipole_fringe;
+    double *R1, *R2, *T1, *T2;
+    double phi, gK;
+    le = atGetDouble(ElemData,"Length"); check_error();
+    bending_angle = atGetDouble(ElemData,"BendingAngle"); check_error();
+    polynom_a = atGetDoubleArray(ElemData,"PolynomA"); check_error();
+    polynom_b = atGetDoubleArray(ElemData,"PolynomB"); check_error();
+    max_order = atGetLong(ElemData, "MaxOrder"); check_error();
+    num_int_steps = atGetLong(ElemData, "NumIntSteps"); check_error();
+    type = atGetLong(ElemData, "Type"); check_error();
+    multipole_fringe = atGetLong(ElemData, "Type"); check_error();
+    phi = atGetDouble(ElemData,"Phi"); check_error();
+    gK = atGetDouble(ElemData,"gK"); check_error();
+    R1=atGetOptionalDoubleArray(ElemData,"R1"); check_error();
+    R2=atGetOptionalDoubleArray(ElemData,"R2"); check_error();
+    T1=atGetOptionalDoubleArray(ElemData,"T1"); check_error();
+    T2=atGetOptionalDoubleArray(ElemData,"T2"); check_error();
+    ExactHamiltonianPass(r_in, le, polynom_a, polynom_b, T1, T2, R1, R2, max_order,
+             num_int_steps, phi, type, gK, multipole_fringe, num_particles);
+    Elem = (struct elem*)atMalloc(sizeof(struct elem));
```

Comment:
The `Elem` structure is intended to store the elements parameters on the 1st pass, and be reused afterwards without needing access to the Python or Matlab variables. Here you store an incomplete structure (`Type`, `gK`, `MultipleFringe` missing), and you don't use it on the next calls, slowing down the tracking by always accessing directly the Matlab or python attributes.